benchmarks/monitoring_log_benchmark.ipynb

```diff
@@ -0,0 +1,227 @@
+    "            y = df['cores'],\n",
+    "            name = monitor + ' core usage (G)',\n",
+    "            legendgroup =  'core usage (G)',\n",
+    "        )\n",
+    "\n",
+    "        # Memory Usage trace (Note the legend group)\n",
+    "        trace2 = go.Scatter(\n",
+    "            x = df.time,\n",
+    "            y = df['memory usage (G)'],\n",
+    "            name = monitor + ' memory usage (G)',\n",
+    "            legendgroup =  'memory usage (G)',\n",
+    "        )\n",
+    "\n",
+    "        # Disk Usage trace (Note the legend group)\n",
+    "        trace3 = go.Scatter(\n",
+    "            x = df.time,\n",
+    "            y = df['disk usage (G)'],\n",
+    "            name = monitor + ' disk usage (G)',\n",
+    "            legendgroup =  'disk usage (G)',\n",
+    "        )\n",
+    "        \n",
+    "        # Add to data collector for plotly\n",
+    "        data = data + [trace1,trace2,trace3]\n",
+    "        \n",
+    "        # get the earliest time (aka start of task) and add it to early collector\n",
+    "        earlys = earlys + [min(time)]\n",
+    "    # TODO figure out a better way to make sure these annotations, which can be long, do not overlap\n",
+    "    def height(num, x = 40, maxh = 60, by = 10):\n",
+    "        \"\"\"Generator to alternate heights of annotation labels\n",
+    "            Inputs:\n",
+    "                num: number of heights we want to output\n",
+    "                minh: min height of annotation\n",
+    "                maxh: max height of annotation\n",
+    "                by: amount to increase height every time\n",
+    "            Outputs:\n",
+    "                Yields varying heights from minh, minh + by... maxh, minh, minh + by... \n",
+    "        \"\"\"\n",
+    "        n = 0\n",
+    "        minh = 40\n",
+    "        while n < num:\n",
+    "            if minh == maxh:\n",
+    "                yield x\n",
+    "                minh = 40\n",
+    "                n +=1 \n",
+    "            yield minh\n",
+    "            minh += by\n",
+    "            n += 1\n",
+    "    # get heights of annotations\n",
+    "    heights = height(len(events))\n",
+    "    # generate layout using collectors\n",
+    "    layout = dict(title = title,\n",
+    "                  xaxis = dict(title = 'Time', range = [min(earlys), min(earlys)+ timedelta(hours=hours)]),\n",
+    "                  yaxis = dict(title = '% Usage', range = [0,100]),\n",
+    "                 annotations=\n",
+    "                    list(map(lambda h : dict(\n",
+    "                        x = h[0],\n",
+    "                        y = next(heights),\n",
+    "                        xref = 'x',\n",
+    "                        yref = 'y',\n",
+    "                        text = h[1],\n",
+    "                        textangle = -45,\n",
+    "                        showarrow = False), sorted(events, key=lambda x : x[0]))))\n",
```

Comment:
This assignment seems convoluted.  A few small, simple statements are better than one large, complicated statement.  Please:

* Use dictionary and list literals where possible; it's more conventional.
* Extract the code creating the `annotations` value into separate variables
  * Avoid multi-line lambdas
  * Assign the dictionary in your map's `func` argument in a separate variable
  * Assign the list in your map's `iter` argument in a separate variable

Reply:
Fixed